cmb2-flexible-content-field.php

```diff
@@ -0,0 +1,447 @@
+									$saved[ $i ][ $grouped_data['supporting_field_id'] ][] = $grouped_data['supporting_field_value'];
+									// Reset var to the actual value.
+									$_new_val[ $group_index ] = $grouped_data['value'];
+								}
+								$new_val = $_new_val;
+							} else {
+								// Add the supporting data to the $saved array stack.
+								$saved[ $i ][ $new_val['supporting_field_id'] ] = $new_val['supporting_field_value'];
+								// Reset var to the actual value.
+								$new_val = $new_val['value'];
+							}
+						}
+
+						$saved[ $i ][ $sub_id ] = $new_val;
+					}
+				}
+				$saved[ $i ] = CMB2_Utils::filter_empty( $saved[ $i ] );
+			}
+			$saved = CMB2_Utils::filter_empty( $saved );
+
+			return $saved;
+		}
+
+
+		/**
+		 * Add Flexible content scripts and styles
+		 */
+		public function add_scripts() {
+			wp_enqueue_script( 'cmb2-flexible-content', plugin_dir_url( __FILE__ ) . 'assets/js/cmb2-flexible.js', array( 'jquery', 'cmb2-scripts' ), '0.1', true );
+			wp_enqueue_style( 'cmb2-flexible-styles', plugin_dir_url( __FILE__ ) . 'assets/css/cmb2-flexible.css', array( 'cmb2-styles' ), '0.1' );
+		}
+
+		/**
+		 * Handle AJAX request for a new flexible row
+		 *
+		 * Creates a new group based on a few variables, renders the output
+		 * then returns it.
+		 */
+		public function handle_ajax() {
+			if ( ! ( isset( $_POST['cmb2_ajax_nonce'] ) && wp_verify_nonce( $_POST['cmb2_ajax_nonce'], 'ajax_nonce' ) ) ) {
+				die();
+			}
+
+			$type = isset( $_POST['type'] ) ? sanitize_key( wp_unslash( $_POST['type'] ) ) : ''; // Input var okay.
+			$metabox_id = isset( $_POST['metabox_id'] ) ? sanitize_key( wp_unslash( $_POST['metabox_id'] ) ) : ''; // Input var okay.
+			$field_id = isset( $_POST['field_id'] ) ? sanitize_key( wp_unslash( $_POST['field_id'] ) ) : ''; // Input var okay.
+			$index = isset( $_POST['latest_index'] ) ? absint( $_POST['latest_index'] ) + 1 : 0; // Input var okay.
+
+			$metabox = cmb2_get_metabox( $metabox_id );
+			$field = cmb2_get_field( $metabox_id, $field_id );
+			$group = $this->create_group( $type, $field, $index );
+
+			ob_start();
+			$this->render_group( $metabox, $field->_id(), $group, $type );
+			$output = ob_get_clean();
+
+			wp_send_json_success( $output );
+		}
+
+		/**
+		 * Render group field wrapped in a flexible container.
+		 *
+		 * If override is set to true, also add meta value override filter
+		 *
+		 * @param  object  $metabox  Metabox instance.
+		 * @param  object  $group    Group field instance.
+		 * @param  string  $type     Layout type.
+		 * @param  boolean $override Should the layout field be added.
+		 */
+		public function render_group( $metabox, $field_id, $group, $type, $override = false ) {
+			$group_name = $group->_id();
+			$index = $group->args['array_key'];
+			$field_key = $field_id . '[' . $index . ']';
+
+			echo '<div class="cmb-row cmb-flexible-row" data-groupid="' . esc_attr( $group_name ) . '" data-groupindex="' . absint( $index ) . '">';
```

Comment:
I'd prefer it if we could break out of PHP here, rather than repeating echo after echo. 🌴